Add support for Raspberry Pi 5

dpoulios opened the pull request:
- Added basic support for rpi5 (mmc, usb, ethernet, devicetree)
- Drivers ported from the oficial Raspberry Pi kernel (https://github.com/raspberrypi/linux/) using commit da87f91ad8450ccc5274cd7b6ba8d823b396c96f
- TODO: Fix random mac address on ethernet

rouming commented:
Thanks @dpoulios for your contribution. A few questions and requests from my side, until @rene ramps up for review.
1. Do we need a special rp5 branch or we can safely merge to the arm64 generic? Can we break something?
2. Is it possible to split your long commit on a separate set of commits which were cherry-picked from the RP branch keeping them untouched (keeping authors and having original sha in the commit message). This will make review easier and atomic.

dpoulios commented:
> Thanks @dpoulios for your contribution. A few questions and requests from my side, until @rene ramps up for review.
> 
>     1. Do we need a special rp5 branch or we can safely merge to the arm64 generic? Can we break something?
> 
>     2. Is it possible to split your long commit on a separate set of commits which were cherry-picked from the RP branch keeping them untouched (keeping authors and having original sha in the commit message). This will make review easier and atomic.

1. I don't think we are breaking anything, but I believe it's better to have some testing on other boards with similar hardware first (besides rpi5 testing), to ensure everything is still running properly. I have already tested it on rpi4, which is probably the most relevant board driver-wise and it seems to be ok.
2. This is not so straightforward, because a) we would probably have to cherry-pick many commits for each single driver, with some of them also having irrelevant code and b) due to the fact that the raspberry kernel is more recent (6.6), I had to make modifications on some cases, to make them build and run properly. I could have a check, however, to see how much effort is needed for this, if we decide it is a better approach.

rene commented:
> > Thanks @dpoulios for your contribution. A few questions and requests from my side, until @rene ramps up for review.
> > ```
> > 1. Do we need a special rp5 branch or we can safely merge to the arm64 generic? Can we break something?
> > 
> > 2. Is it possible to split your long commit on a separate set of commits which were cherry-picked from the RP branch keeping them untouched (keeping authors and having original sha in the commit message). This will make review easier and atomic.
> > ```
> 
>     1. I don't think we are breaking anything, but I believe it's better to have some testing on other boards with similar hardware first (besides rpi5 testing), to ensure everything is still running properly. I have already tested it on rpi4, which is probably the most relevant board driver-wise and it seems to be ok.
> 
>     2. This is not so straightforward, because a) we would probably have to cherry-pick many commits for each single driver, with some of them also having irrelevant code and b) due to the fact that the raspberry kernel is more recent (6.6), I had to make modifications on some cases, to make them build and run properly. I could have a check, however, to see how much effort is needed for this, if we decide it is a better approach.

@dpoulios , break your changes into multiple commits it's a better approach. I understand the pitfalls of cherry-picking every single commit from RPi's kernel, I don't think you need to do that, but it's really important to clarify whether it was needed to introduce changes in the ported code or not, we have ported a lot of drivers in our kernel, here are some commits as an example:

https://github.com/lf-edge/eve-kernel/commit/e20b2354baaf3cff9a1f2ece699087e15b25cc36
https://github.com/lf-edge/eve-kernel/commit/e63c784657ecf20172f831b3c6a9e9d4ddce676d
https://github.com/lf-edge/eve-kernel/commit/42d2c10264f9376c65149cb0647ccb81c35731ad

Multiple commits also helps us to use `git bisect` to address eventual issues... you didn't port too much files in this case, but I would expect at least device tree files, configs, codes that you changed being introduced into individual commits....

rene commented on arch/arm64/configs/eve_defconfig:
```diff
@@ -731,6 +735,9 @@ CONFIG_PINCTRL_IMX8QM=y
 CONFIG_PINCTRL_IMX8QXP=y
 CONFIG_PINCTRL_IMX8DXL=y
 CONFIG_PINCTRL_SPRD_SC9860=y
+CONFIG_PINCTRL_BCM2712=y
```
It's really needed to built-in all these configs? Our kernel it's an ARM64 generic kernel that can run on other devices as well, so if there is something that can be built as a module, would be better to let as a module...

rene commented on arch/arm64/configs/eve_defconfig:
```diff
@@ -555,6 +555,9 @@ CONFIG_HNS3=y
 CONFIG_HNS3_ENET=y
 CONFIG_NET_VENDOR_MEDIATEK=y
+CONFIG_NET_VENDOR_CADENCE=y
+CONFIG_MACB=y
```
Please, after adding your changes to the eve_defconfig file, run `make eve_defconfig && make savedefconfig` and replace the `eve_defconfig` file with the generated `defconfig`, in this way we always keep only the needed set of options in the default config file and it's easier to merge the changes with other variants of our kernel as well...

rene commented on arch/arm/boot/dts/broadcom/bcm2712-rpi.dtsi:
```diff
@@ -0,0 +1,343 @@
+		status = "okay";
+	};
+
+	rpi_rtc: rpi_rtc {
```
I didn't find the driver for the RTC... it was ported as well?

dpoulios commented:
> > > Thanks @dpoulios for your contribution. A few questions and requests from my side, until @rene ramps up for review.
> > > ```
> > > 1. Do we need a special rp5 branch or we can safely merge to the arm64 generic? Can we break something?
> > > 
> > > 2. Is it possible to split your long commit on a separate set of commits which were cherry-picked from the RP branch keeping them untouched (keeping authors and having original sha in the commit message). This will make review easier and atomic.
> > > ```
> > 
> > 
> > ```
> > 1. I don't think we are breaking anything, but I believe it's better to have some testing on other boards with similar hardware first (besides rpi5 testing), to ensure everything is still running properly. I have already tested it on rpi4, which is probably the most relevant board driver-wise and it seems to be ok.
> > 
> > 2. This is not so straightforward, because a) we would probably have to cherry-pick many commits for each single driver, with some of them also having irrelevant code and b) due to the fact that the raspberry kernel is more recent (6.6), I had to make modifications on some cases, to make them build and run properly. I could have a check, however, to see how much effort is needed for this, if we decide it is a better approach.
> > ```
> 
> @dpoulios , break your changes into multiple commits it's a better approach. I understand the pitfalls of cherry-picking every single commit from RPi's kernel, I don't think you need to do that, but it's really important to clarify whether it was needed to introduce changes in the ported code or not, we have ported a lot of drivers in our kernel, here are some commits as an example:
> 
> [e20b235](https://github.com/lf-edge/eve-kernel/commit/e20b2354baaf3cff9a1f2ece699087e15b25cc36) [e63c784](https://github.com/lf-edge/eve-kernel/commit/e63c784657ecf20172f831b3c6a9e9d4ddce676d) [42d2c10](https://github.com/lf-edge/eve-kernel/commit/42d2c10264f9376c65149cb0647ccb81c35731ad)
> 
> Multiple commits also helps us to use `git bisect` to address eventual issues... you didn't port too much files in this case, but I would expect at least device tree files, configs, codes that you changed being introduced into individual commits....

Hello @rene ,

thank you for your comments. I am currently reworking the pull request to split it in multiple commits, each of them containing drivers needed for each basic functionality (mmc, usb, ethernet etc.) and probably one extra commit for common drivers that are dependency of multiple drivers. Is this ok?

I will also try to have as much as possible compiled as modules to keep the kernel more lightweight and finally cleanup _eve_defconfig_.


> I didn't find the driver for the RTC... it was ported as well?

No, the RTC driver has not been ported yet. This pull request includes a first attempt to boot eve on rpi5 and onboard it. Nevertheless, I don't think the RTC driver needs much effort to port it. Do you think I should include it, as well?

dpoulios committed:
Add general support for Raspberry Pi 5

This commit contains drivers that provide basic functionality for RPI5:

- DMA
- MSI-X
- WDT
- Power domains

Ported from:

Repository: https://github.com/raspberrypi/linux
Branch:     rpi-6.6.y
Revision:   da87f91ad8450ccc5274cd7b6ba8d823b396c96f

Signed-off-by: Dimitrios Poulios <dpoulios85@gmail.com>

dpoulios committed:
Add sd card support for Raspberry Pi 5

- Enables sdhci support for RPI5 (sdhci-brcmstb driver)
- Broadcom BCM2712 PINCONF driver (pinctrl-bcm2712) is required

Ported from:

Repository: https://github.com/raspberrypi/linux/
Branch:     rpi-6.6.y
Revision:   da87f91ad8450ccc5274cd7b6ba8d823b396c96f

Signed-off-by: Dimitrios Poulios <dpoulios85@gmail.com>

dpoulios committed:
Add pci/usb support for Raspberry Pi 5

- Updates pcie-brcmstb driver to enable PCI on RPI5
- Adds USB support by porting RP1 MFD & Pinctrl drivers

Ported from:

Repository: https://github.com/raspberrypi/linux/
Branch:     rpi-6.6.y
Revision:   da87f91ad8450ccc5274cd7b6ba8d823b396c96f

Signed-off-by: Dimitrios Poulios <dpoulios85@gmail.com>

dpoulios committed:
Add ethernet support for Raspberry Pi 5

- Updates Cadence MACB driver to enable ethernet on RPI5
- Needs RP1 common clock framework support (clk-rp1 driver)

Ported from:

Repository: https://github.com/raspberrypi/linux/
Branch:     rpi-6.6.y
Revision:   da87f91ad8450ccc5274cd7b6ba8d823b396c96f

Signed-off-by: Dimitrios Poulios <dpoulios85@gmail.com>

dpoulios committed:
dts: Add device tree for Raspberry Pi 5

Ported from:

Repository: https://github.com/raspberrypi/linux/
Branch:     rpi-6.6.y
Revision:   da87f91ad8450ccc5274cd7b6ba8d823b396c96f

Signed-off-by: Dimitrios Poulios <dpoulios85@gmail.com>

dpoulios commented:
- The PR was modified and is now split in multiple commits, each of them containing more information.
- I also did some cleanup, removing portions of the patched code that were not needed.
- eve_defconfig is now clean and auto-generated. Unfortunately, none of the new drivers could be added as module, even though I tried.
- Can be tested on RPI5 with a patch on eve (upcoming pull request on that repo), which updates uboot and modifies grub configuration to support the board.

rene commented:
Thanks again for the PR @dpoulios , the commits look much better. Regarding the RTC driver, if it's not a big deal to port, please, do the port as well, we do rely on hardware RTC to get the first time/date at the boot and avoid certificate issues when trying to communicate with the cloud controller. It's not mandatory tough, but really a must have if the hardware has the RTC, like the RPi 5.

dpoulios committed:
drivers: rtc: add rpi-rtc driver for Raspberry Pi 5

Ported from:

Repository: https://github.com/raspberrypi/linux/
Branch:     rpi-6.6.y
Revision:   da87f91ad8450ccc5274cd7b6ba8d823b396c96f

Signed-off-by: Dimitrios Poulios <dpoulios85@gmail.com>

dpoulios commented:
> Thanks again for the PR @dpoulios , the commits look much better. Regarding the RTC driver, if it's not a big deal to port, please, do the port as well, we do rely on hardware RTC to get the first time/date at the boot and avoid certificate issues when trying to communicate with the cloud controller. It's not mandatory tough, but really a must have if the hardware has the RTC, like the RPi 5.

RTC driver ported & tested

rene commented:
@dpoulios , LGTM, now these changes must be tested with other ARM devices we support (for this case, RPi 4B and iMX8MP devices) to ensure nothing got broken.... I'm looking forward for your PR on EVE introducing the other changes needed for RPi 5. You can also create a DRAFT PR while these changes are not merged, so we can start to review.....

dpoulios commented:
> @dpoulios , LGTM, now these changes must be tested with other ARM devices we support (for this case, RPi 4B and iMX8MP devices) to ensure nothing got broken.... I'm looking forward for your PR on EVE introducing the other changes needed for RPi 5. You can also create a DRAFT PR while these changes are not merged, so we can start to review.....

Hello!

I agree that both eve-kernel & eve changes must be tested on other arm boards. I haven't made the PR for eve yet, because I guess the kernel tag should be first updated there (after this is tested and merged). The commits, however, are already pushed in my eve repo (https://github.com/dpoulios/eve/ - raspberrypi5 branch), so they are ready to be tested, manually handling the kernel tag.

For RPI5, one thing not available yet is the hdmi output of grub (only shows up in uart), so we don't have the boot options menu. I will fix this in the near future.

rene commented:
@dpoulios, I've tested this PR with the following devices:

1. Raspberry Pi 4B
2. Raspberry Pi 5
3. Advantech EPC-R3720 (iMX.8MP)

All devices are working properly, we are ready to merge this PR. Thank you for your contribution.
